### PR TITLE
Intervals are updated properly for Unnest queries

### PR DESCRIPTION
**Pull request by somu-imply**
Previously unnest queries were not updated with the correct intervals. For example 
if I run a query as
```
select * from foo
, UNNEST(MV_TO_ARRAY("dim3")) as ud(d3) 
where __time >= TIMESTAMP '2000-01-02 00:00:00' and __time <= TIMESTAMP '2000-01-03 00:10:00' 
```
The following plan shows up
```
{
  "queryType": "scan",
  "dataSource": {
    "type": "unnest",
    "base": {
      "type": "filter",
      "base": {
        "type": "table",
        "name": "foo"
      },
      "filter": {
        "type": "range",
        "column": "__time",
        "matchValueType": "LONG",
        "lower": 946771200000,
        "upper": 946858200000
      }
    },
    "virtualColumn": {
      "type": "expression",
      "name": "j0.unnest",
      "expression": "\"dim3\"",
      "outputType": "STRING"
    },
    "unnestFilter": null
  },
  "intervals": {
    "type": "intervals",
    "intervals": [
      "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"
    ]
  },
  "resultFormat": "compactedList",
  "limit": 1001,
  "columns": [
    "__time",
    "dim1",
    "dim2",
    "dim3",
    "j0.unnest",
    "m1",
    "m2"
  ],
  "legacy": false,
  "context": {
    "enableUnnest": true,
    "queryId": "a6147ad4-13c4-4c9c-9213-42e3d808d00f",
    "sqlOuterLimit": 1001,
    "sqlQueryId": "a6147ad4-13c4-4c9c-9213-42e3d808d00f",
    "useNativeQueryExplain": true
  },
  "granularity": {
    "type": "all"
  }
}
```

The actual interval should have been
```
"intervals": {
    "type": "intervals",
    "intervals": [
      "2000-01-02T00:00:00.000Z/2000-01-03T00:10:00.001Z"
    ]
  }
```
With this change, we have added unit tests to ensure the intervals are correctly updated 
 
This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 1]
```diff
@@ -788,6 +790,24 @@ static Pair<DataSource, Filtration> getFiltration(
   {
     if (!canUseIntervalFiltering(dataSource)) {
       return Pair.of(dataSource, toFiltration(filter, virtualColumnRegistry.getFullRowSignature(), false));
+    } else if (dataSource instanceof UnnestDataSource) {
+      // UnnestDataSource can have another unnest data source
+      // join datasource, filtered data source, etc as base
+      Pair<DataSource, Filtration> pair = getFiltration(
+          ((UnnestDataSource) dataSource).getBase(),
```
Comment (pranavbhole) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:797]: Can you please add test case for nested UnnestDatasources ? just want to make sure that correct interval populates correctly for 2-3 nested case.

Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:797]: Added more tests with 2 and 3 levels of nesting

Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:797]: We are not checking if the underlying data source can use interval filtering. What's the behavior when we replace the table data source with something like a lookup and do an unnest on that? We shouldn't call `optimize` in that case since lookups can't use interval filtering.

Why is the `UNNEST` datasource not triggering the final `else` branch in this scenario if the underlying table `canUseIntervalFiltering` and the very first if branch if the data source can't?

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 2]
```diff
@@ -788,6 +790,24 @@ static Pair<DataSource, Filtration> getFiltration(
...
+          filter,
+          virtualColumnRegistry,
+          joinableFactoryWrapper
+      );
+      return Pair.of(dataSource, pair.rhs);
+    } else if (dataSource instanceof FilteredDataSource) {
+      final FilteredDataSource filteredDataSource = (FilteredDataSource) dataSource;
+      final Filtration baseFiltration = Filtration.create(filteredDataSource.getFilter())
+                                                  .optimize(virtualColumnRegistry.getFullRowSignature());
```
Comment (pranavbhole) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:806]: should we optimize optimizeFilterOnly if it is not using the intervals?

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java (hunk 1 above)
Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:797]: We have pushed the filter into the filteredDataSource, so if the last else is triggered, the filter is null and the filtration is not applied and query segment spec of eternity takes over. Regarding the comment on datasource using interval filtering, I'll add a case under unnest to check if the base data source can use interval filtering and trigger the same logic as the last else

Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:797]: in the original code before the unnest handling this would appear as
```
datasource = UnnestDS(FilteredDS(foo,timeFilter))
filter = null
```

The presence of a filter would always trigger a filter data source. So that flow was not able to handle this without the special handling for UnnestDataSource which just delegates to its children and the Filter data source which uses the filter present with it

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java (hunk 2 above)
Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:806]: This could be replaced by a call as follows 
```
final boolean status = canUseIntervalFiltering(filteredDataSource);
 final Filtration baseFiltration = toFiltration(filteredDataSource.getFilter(), virtualColumnRegistry.getFullRowSignature(), status);
      
```

Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java]: Updated this

**Review by pranavbhole — APPROVED** [review]

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 3]
```diff
@@ -788,6 +790,28 @@ static Pair<DataSource, Filtration> getFiltration(
...
+      final FilteredDataSource filteredDataSource = (FilteredDataSource) dataSource;
+      final boolean useIntervalFiltering = canUseIntervalFiltering(filteredDataSource);
+      final Filtration baseFiltration = toFiltration(
+          filteredDataSource.getFilter(),
+          virtualColumnRegistry.getFullRowSignature(),
+          useIntervalFiltering
+      );
+      // Adds the intervals from the filter of filtered data source to query filtration
+      final Filtration queryFiltration = Filtration.create(filter, baseFiltration.getIntervals())
+                                                   .optimize(virtualColumnRegistry.getFullRowSignature());
```
Comment (LakshSingla) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:813]: `toFiltration` in the above call is already optimizing. Why do we require this again?

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 4]
```diff
@@ -788,6 +790,28 @@ static Pair<DataSource, Filtration> getFiltration(
   {
     if (!canUseIntervalFiltering(dataSource)) {
       return Pair.of(dataSource, toFiltration(filter, virtualColumnRegistry.getFullRowSignature(), false));
+    } else if (dataSource instanceof UnnestDataSource) {
```
Comment (LakshSingla) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:793]: nit: It won't make a difference with the logic as is, but let's put this branch at the top since it is the most specific of the branches.

**Review by LakshSingla — COMMENTED** [review]
Let us add some comments and assumptions in the `getFiltration` method's Javadoc:

1. FilteredDataSource is always on the UnnestDataSource
2. UnnestDataSource's analysis gives the base data source which cannot have filters. 

Also, please add why this special handling is required. 
We should add a test in the query's test which filters on unnest on top of a lookup.

**Comment by LakshSingla** [comment]
I think there are cases when there are multiple `FilteredDataSources` stacked on top of each other directly or indirectly. 
For example -> FilteredDS -> UnnestDS -> FilteredDS -> UnnestDS.
How would this code handle those cases, will it ignore the second UnnestedDS?

What if the base table of the above chain has a filter of itself? (Is it possible in the first place?) If not then we should clearly highlight that assumption in the getFiltration method. Otherwise, it would be good if the code could handle cases like these automatically. 

Also, UNNEST on MSQ got merged so let's make sure that the tests pass with them and we prune the segments properly.

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java (hunk 3 above)
Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:813]: The `toFiltration` call creates the filtration object from the filter inside the filteredDS. The second call adds the intervals created in this step to the filter which is passed to the initial call of the `getFiltration` method

**Comment by somu-imply** [comment]
A filteredDS is created only inside the rel for Unnest, ensuring it only grabs the outermost filter and, if possible, pushes it down inside the data source. So a chain of Filter->Unnest->Filter is typically impossible when the query is done through SQL. Also, Calcite has filter reduction rules that push filters deep into base data sources for better query planning. Hence the case that can be seen is a bunch of unnests followed by a terminal filteredDS like Unnest->Unnest->FilteredDS. I have UTs resembling the same. A base table with a chain of filters is synonymous with a filteredDS. In case there are filters present in the getFiltration call we still update the interval by: 1) creating a filtration from the filteredDS's filter and 2) Updating the interval of the outer filter with the intervals in step 1. 

I'll merge the updated master into this branch to ensure that the tests pass correctly after this change.

File: extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java [hunk 5]
```diff
@@ -1798,7 +1798,7 @@ public void testTimeColumnAggregationFromExtern() throws IOException
         .setExpectedValidationErrorMatcher(
             new DruidExceptionMatcher(DruidException.Persona.ADMIN, DruidException.Category.INVALID_INPUT, "general")
                 .expectMessageIs(
-                    "Query planning failed for unknown reason, our best guess is this "
+                    "Query could not be planned. A possible reason is "
```
Comment (somu-imply) [inline extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/MSQSelectTest.java:1801]: Nit. this is a cosmetic change, it was too small to open a separate PR for this so just adding it here

**Comment by somu-imply** [comment]
A defensive check was added, code was slightly reformatted along with a comment

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 6]
```diff
@@ -786,8 +789,44 @@ static Pair<DataSource, Filtration> getFiltration(
...
+      // A base table with a chain of filters is synonymous with a filteredDS.
+      // In case there are filters present in the getFiltration call we still update the interval by:
+      // 1) creating a filtration from the filteredDS's filter and
+      // 2) Updating the interval of the outer filter with the intervals in step 1, and you'll see these 2 calls in the code
+      final FilteredDataSource filteredDataSource = (FilteredDataSource) dataSource;
+      // Defensive check as in the base of a filter cannot be another filter
+      final DataSource baseOfFilterDataSource = filteredDataSource.getBase();
+      if (baseOfFilterDataSource instanceof FilteredDataSource) {
+        throw DruidException.defensive("Cannot create a filteredDataSource using another filteredDataSource as a base");
+      }
```
Comment (LakshSingla) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:819]: Instead of getBase(), we need to do a recursive check on the `filteredDataSource.getChildren()`.
i.e. FilteredDS -> UnnestDS -> FilteredDS is disallowed ❌ .
We are only accounting for the top-level filtered DS, therefore any intervals in any nested DS.

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 7]
```diff
@@ -786,8 +789,44 @@ static Pair<DataSource, Filtration> getFiltration(
...
+    } else if (dataSource instanceof FilteredDataSource) {
+      // A filteredDS is created only inside the rel for Unnest, ensuring it only grabs the outermost filter
+      // and, if possible, pushes it down inside the data source.
+      // So a chain of Filter->Unnest->Filter is typically impossible when the query is done through SQL.
+      // Also, Calcite has filter reduction rules that push filters deep into base data sources for better query planning.
+      // Hence, the case that can be seen is a bunch of unnests followed by a terminal filteredDS like Unnest->Unnest->FilteredDS.
+      // A base table with a chain of filters is synonymous with a filteredDS.
+      // In case there are filters present in the getFiltration call we still update the interval by:
+      // 1) creating a filtration from the filteredDS's filter and
+      // 2) Updating the interval of the outer filter with the intervals in step 1, and you'll see these 2 calls in the code
```
Comment (LakshSingla) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:813]: nit: push this branch above due to the same reason, perhaps we can move it to another PR if there are no further comments.

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java (hunk 6 above)
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:819]: Rethinking about it, I see that we are only handling the cases where UnnestDS is at the top level. A query chain like QueryDS -> UnnestDS -> FilteredDS -> .... won't set the intervals specified in the FilteredDS properly.

Comment (somu-imply) [reply sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:819]: The QueryDS would be handled separately where it would find its own UnnestDS and FilteredDS and set the interval for the inner subquery. The outer query should still be eternity. This is similar to how Druid handles scan over queryDS. I'll add the examples. I agree with the other comment that we need to recursively for the case of Unnest->Filter->Unnest->Filter. I'll update the comment adding that a chain cannot start with a filterDS

**Review by clintropolis — APPROVED** [review]

**Review by LakshSingla — APPROVED** [review]
LGTM! Thanks for the PR, and for entertaining the questions I had around UNNEST.
